Timeline density improvements and refactoring

AaronPlave opened the pull request:
Improves the density of the timeline through re-organization of the layout and the addition of various new features. 

<img width="1434" alt="image" src="https://github.com/NASA-AMMOS/aerie-ui/assets/4419607/934017e4-8f5d-43d7-a437-7724104c8346">

<br/><br/>

Changes:
- Display plan time bounds in DOY format to the left of the timeline histogram.
- Refactor timeline x axis time visualization to display dynamically formatted UTC and user local time.
- Display row header and controls in the left column of the row instead of horizontally stacking the header.
- Resize row header (i.e. marginLeft in the timeline view settings, TODO rename this?) by clicking and dragging on the right side of the row header column.
- Move row drag handle to the middle of uncollapsed row headers, allow users to also drag rows from the row title.
- Row option to show only activity directives.
- Improved row dragging styles.
- Resolves existing row jitter bug @bradNASA and @ewferg discovered (no issue for this?) which was primarily caused by the interaction between row drag-n-drop and row resize. Spatially separating these functionalities eliminated this issue though keep an eye out.
- Render line and x-range plots when rows are collapsed.
- Display heatmap of directives and/or spans for collapsed activity layers.
- Change row y-axis ticks to use a dotted line.
- Move y-axes to within the row header and separate them using vertical lines
- Enforce the user's tick count number as defined in the view instead of using the normal d3 method which only takes tick count as a suggestion.
- Color y-axis tick values by resource color if a single resource is associated with the axis.
- Three dot menu in row header with row actions (Edit Row, Move Up/Down, Duplicate Row, Delete Row, Show Simulated Activities/Directives/Both, Insert Row).
- Right click row header context menu with row actions.
- Display resource names under row titles and color using layer color.
- Format y axis tick values using SI-prefixes

Other TODOs:
- [x] Fix alignment of hover time cursor
- [x] Fix box shadow styling issue with first row when editing
- [x] Bug: duplicated row not initially showing up in timeline editor
- [x] Add optional name field to layers and use it for line layer labeling in row header if supplied
- [x] Show x-range name in row header
- [x] Add tooltip to resource names in row header, allow users to copy resource names if overridden/truncated
- [x] Potentially switch back to d3 default axes for pretty numbers
- [x] Add duplicate to main timeline row buttons

Closes #954
Closes #443
Closes #453
Closes #746

AaronPlave committed:
Timeline density improvements and refactoring WIP

AaronPlave committed:
Clean

AaronPlave committed:
Fixes

AaronPlave committed:
Color y axis labels if appropriate

jeffpamer commented:
![image](https://github.com/NASA-AMMOS/aerie-ui/assets/888818/1a801219-efa4-4c13-89bc-191c29d1b660)

Looks like there's a z-indexing issue with the row settings dialog.

AaronPlave commented:
> ![image](https://user-images.githubusercontent.com/888818/280129942-1a801219-efa4-4c13-89bc-191c29d1b660.png)
> 
> Looks like there's a z-indexing issue with the row settings dialog.

Yeah, fixing that right now by moving all of that functionality into the TimelineContextMenu.

AaronPlave committed:
Fixes, WIP overhaul of context menu usage in row header

AaronPlave committed:
Add close on escape behavior to context menu

AaronPlave committed:
Fix for hz guide rendering

AaronPlave committed:
Duplicate row functionality

AaronPlave committed:
Fixes and ability to right click anywhere in row header

AaronPlave committed:
Cursor alignment fixes

AaronPlave committed:
Style fix

AaronPlave committed:
Improve x time label formatting, add more d3 tick ranges to prevent tick overlap

AaronPlave committed:
Fix active row state styling

AaronPlave committed:
Use units in row resource labels

AaronPlave committed:
Style fix

AaronPlave committed:
Clean

AaronPlave committed:
Fixes

duranb reviewed (commented):
Some initial findings:
1. When the activity layer is collapsed so that the heatmap is shown, any activities right at the beginning of the plan start don't seem to be rendered or are being covered up by the leftmost part of the container. Even moving the simulation start guideline doesn't reveal them.
2. Dragging rows using the title is only on the text. Should it be extended for the whole horizontal row that the title is in?

AaronPlave committed:
Fix for row component sizing to ensure left side is visible

AaronPlave commented:
> Some initial findings:
> 
> 1. When the activity layer is collapsed so that the heatmap is shown, any activities right at the beginning of the plan start don't seem to be rendered or are being covered up by the leftmost part of the container. Even moving the simulation start guideline doesn't reveal them.
> 2. Dragging rows using the title is only on the text. Should it be extended for the whole horizontal row that the title is in?

Resolved 1).

Regarding 2) – Right, can only drag from title and from drag handle that shows up on the left side of the row header on row hover. Issue I was facing was conflict between row drag and row height resizing which was causing the existing jitter bug where the handlers would step on each other pretty badly. Didn't manage to find a way around that issue so figured we could circumvent it by not having the two functionalities exist in the same space. That said if you find a way to make them exist in harmony I think it'd definitely be nicer to have the whole thing be draggable.

AaronPlave commented:
> > Some initial findings:
> > 
> > 1. When the activity layer is collapsed so that the heatmap is shown, any activities right at the beginning of the plan start don't seem to be rendered or are being covered up by the leftmost part of the container. Even moving the simulation start guideline doesn't reveal them.
> > 2. Dragging rows using the title is only on the text. Should it be extended for the whole horizontal row that the title is in?
> 
> Resolved 1).
> 
> Regarding 2) – Right, can only drag from title and from drag handle that shows up on the left side of the row header on row hover. Issue I was facing was conflict between row drag and row height resizing which was causing the existing jitter bug where the handlers would step on each other pretty badly. Didn't manage to find a way around that issue so figured we could circumvent it by not having the two functionalities exist in the same space. That said if you find a way to make them exist in harmony I think it'd definitely be nicer to have the whole thing be draggable.

2 is now resolved well enough - agreed on making title area as wide as it'll go to expand general drag area.

AaronPlave committed:
Remove unused files

AaronPlave committed:
Add layer name property and ability to override it for resource layers. Layer name overrides resource name in row header if present. Add copy-able tooltip to row header resource label.

AaronPlave committed:
Tweak default y axis name

AaronPlave committed:
Pretty y-axis tick numbers

AaronPlave committed:
Duplicate row from TimelineEditor

AaronPlave committed:
Lint fix

duranb reviewed (approved):
great work @AaronPlave! i love the new design @lklyne!

AaronPlave committed:
Format numbers with SI-prefixes

AaronPlave committed:
Fix resource label tooltip z indexing

dandelany reviewed (approved):
This is looking great @AaronPlave! Tested all of the features mentioned above with no major issues - all of my notes are minor & totally optional nitpicks/improvements:
* Small z-index bug - when you hover over the Y axis legend label on a line chart row AND that row is selected, the row boundary line shows above the tooltip
![image](https://github.com/NASA-AMMOS/aerie-ui/assets/734221/0710f766-a049-4f4a-b650-9d9598cf3278)

* Would be nice to have some better handling for naming copies of copies, by duplicating rows you can easily end up with "My row (copy) (copy) (copy)"
* When dragging row edges to resize, rows have a minimum height and minimum left margin (now the header area). These minimums *aren't* enforced if you edit the height/margin using the number input the Timeline Editor panel, and probably should be - otherwise you can accidentally end up with a very small row
* The new row context menu is nice! It makes me want a slight extension of it - if you right click *below* the last row, it would be nice to have a context menu with a shortcut to add a new row

AaronPlave commented:
> This is looking great @AaronPlave! Tested all of the features mentioned above with no major issues - all of my notes are minor & totally optional nitpicks/improvements:

Fixed, good catch!
> * Small z-index bug - when you hover over the Y axis legend label on a line chart row AND that row is selected, the row boundary line shows above the tooltip

>   ![image](https://user-images.githubusercontent.com/734221/282123779-0710f766-a049-4f4a-b650-9d9598cf3278.png)

Agreed but will punt on this.
> * Would be nice to have some better handling for naming copies of copies, by duplicating rows you can easily end up with "My row (copy) (copy) (copy)"

Implemented.
> * When dragging row edges to resize, rows have a minimum height and minimum left margin (now the header area). These minimums _aren't_ enforced if you edit the height/margin using the number input the Timeline Editor panel, and probably should be - otherwise you can accidentally end up with a very small

One issue here is that if your rows take up more than the available height there will be no additional room below the last row. How would you feel about adding an "Insert row" context menu option that is always available?
> * The new row context menu is nice! It makes me want a slight extension of it - if you right click _below_ the last row, it would be nice to have a context menu with a shortcut to add a new row

dandelany commented:
> How would you feel about adding an "Insert row" context menu option that is always available?

Seems reasonable to me! 👍

AaronPlave committed:
Fixes

AaronPlave committed:
Insert row from context menu, style tweaks

AaronPlave committed:
Add row duplication test
